src/librustc/lint/context.rs

```diff
@@ -1082,6 +1089,112 @@ impl LateLintPass for GatherNodeLevels {
+            match span {
+                Some(span) => sess.span_warn(span, &warning[..]),
+                None => sess.warn(&warning[..]),
+            };
+            CheckLintNameResult::Mentioned
+        },
+        Some(&Removed(ref reason)) => {
+            let warning = format!("lint {} has been removed: {}", lint_name, reason);
+            match span {
+                Some(span) => sess.span_warn(span, &warning[..]),
+                None => sess.warn(&warning[..])
+            }
+            CheckLintNameResult::Mentioned
+        },
+        None => {
+            match lint_cx.lint_groups.get(lint_name) {
+                None => {
+                    CheckLintNameResult::NoLint
+                }
+                Some(_) => {
+                    /* lint group exists */
+                    CheckLintNameResult::Ok
+                }
+            }
+        }
+        Some(_) => {
+            /* lint exists */
+            CheckLintNameResult::Ok
+        }
+    }
+}
+
+// Checks the validity of lint names derived from attributes
+fn check_lint_attribute(cx: &LateContext, attr: &ast::Attribute) {
+    for result in gather_attr(attr) {
+        match result {
+            Err(_) => {
+                // Malformed lint attr. Reported by with_lint_attrs
+                continue;
+            }
+            Ok((lint_name, _, span)) => {
+                match check_lint_name(&cx.tcx.sess, &cx.lints, &lint_name[..], Some(span)) {
+                    CheckLintNameResult::Ok => (),
+                    CheckLintNameResult::Mentioned => (),
+                    CheckLintNameResult::NoLint => {
+                        cx.span_lint(builtin::UNKNOWN_LINTS, span,
+                                     &format!("unknown lint: `{}`",
+                                              lint_name));
+                    }
+                }
+            }
+        }
+    }
+}
+
+// Checks the validity of lint names derived from the command line
+fn check_lint_cmdline_name(sess: &Session, lint_cx: &LintStore,
+                           lint_name: &str, level: Level) {
+    let explain_cmdline;
+
+    match check_lint_name(sess, lint_cx, lint_name, None) {
+        CheckLintNameResult::Ok => explain_cmdline = false,
+        CheckLintNameResult::Mentioned => explain_cmdline = true,
+        CheckLintNameResult::NoLint => {
+            sess.err(&format!("unknown lint: `{}`", lint_name));
+            explain_cmdline = true;
+        }
+    }
+
+    if explain_cmdline {
+        let msg = format!("requested on the command line with `{} {}`",
+                          match level {
+                              Level::Allow => "-A",
+                              Level::Warn => "-W",
+                              Level::Deny => "-D",
+                              Level::Forbid => "-F",
+                          },
+                          lint_name);
+        sess.note_without_error(&msg);
```

Comment:
Ah, yes, this is true. It'd be better to return the error and attach the note to it -- that would also mean that the caller is always the one that ultimately reports all warnings/errors, which seems good.